lib/browser/api/net.ts

```diff
@@ -526,6 +529,122 @@ export function request (options: ClientRequestConstructorOptions | string, call
   return new ClientRequest(options, callback);
 }
 
+function createDeferredPromise<T, E extends Error = Error> (): { promise: Promise<T>; resolve: (x: T) => void; reject: (e: E) => void; } {
+  let res: (x: T) => void;
+  let rej: (e: E) => void;
+  const promise = new Promise<T>((resolve, reject) => {
+    res = resolve;
+    rej = reject;
+  });
+
+  return { promise, resolve: res!, reject: rej! };
+}
+
+export function fetch (input: RequestInfo, init?: RequestInit): Promise<Response> {
+  const p = createDeferredPromise<Response>();
+  let req: Request;
+  try {
+    req = new Request(input, init);
+  } catch (e: any) {
+    p.reject(e);
+    return p.promise;
+  }
+
+  if (req.signal.aborted) {
+    // 1. Abort the fetch() call with p, request, null, and
+    //    requestObject’s signal’s abort reason.
+    const error = (req.signal as any).reason ?? new DOMException('The operation was aborted.', 'AbortError');
+    p.reject(error);
+
+    if (req.body != null && isReadable(req.body as unknown as NodeJS.ReadableStream)) {
+      req.body.cancel(error).catch((err) => {
+        if (err.code === 'ERR_INVALID_STATE') {
+          // Node bug?
+          return;
+        }
+        throw err;
+      });
+    }
+
+    // 2. Return p.
+    return p.promise;
+  }
+
+  let locallyAborted = false;
+  req.signal.addEventListener(
+    'abort',
+    () => {
+      // 1. Set locallyAborted to true.
+      locallyAborted = true;
+
+      // 2. Abort the fetch() call with p, request, responseObject,
+      //    and requestObject’s signal’s abort reason.
+      const error = (req.signal as any).reason ?? new DOMException('The operation was aborted.', 'AbortError');
+      p.reject(error);
+      if (req.body != null && isReadable(req.body as unknown as NodeJS.ReadableStream)) {
+        req.body.cancel(error).catch((err) => {
+          if (err.code === 'ERR_INVALID_STATE') {
+            // Node bug?
+            return;
+          }
+          throw err;
+        });
+      }
+
+      r.abort();
+    },
+    { once: true }
+  );
+
+  const origin = req.headers.get('origin') ?? undefined;
+  // We can't set credentials to same-origin unless there's an origin set.
+  const credentials = req.credentials === 'same-origin' && !origin ? 'include' : req.credentials;
+
+  const r = request({
+    // TODO: session
```

Comment:
Hm, I forgot that I left this outstanding. You can't currently specify a session for the `fetch` request to happen in, so all `fetch` requests will happen in the default session. A couple of options for remedying this:

1. Extend the `RequestInit` type to allow passing in `session` or `partition`. i.e. `fetch('foo', { session: x })`. This is non-standard, and our type for the 2nd argument to net.fetch would become something like `RequestInit & {session?: Session, partition?: string}`. Would possibly require some docs-parser changes.
2. Move `net.fetch` to `Session`, i.e. `session.defaultSession.fetch(...)`. `net.fetch` could be an alias for `session.defaultSession.fetch(...)`.

I think I'm in favor of #2.

Reply:
I'm in favor of 2 as well. And keeping `net.fetch` serves as a good place for API discovery.

Reply:
Done.